### PR TITLE
Retain ref to requests when running ActionFilterChain

### LINKED ISSUE #103952: SecurityActionFilter performs async actions without retaining refs to requests
An `ActionRequest` may hold nontrivial resources, implementing the `RefCounted` interface to ensure its resources are retained until no longer necessary at which point they are released promptly. Our internal APIs are generally refcount-neutral, acquiring refs as needed to keep resources alive beyond the scope of the executing method and releasing the acquired ref when no longer needed. In particular, async code must acquire a ref before going async and then release the ref once the async execution no longer needs it.

Today `SecurityActionFilter` does not respect this refcounting discipline: along some pathways it performs async actions without first acquiring a ref to the request. By the time the async action comes to being executed the request may already be closed.

This is not technically a bug today because in practice no `ActionRequest` instances hold nontrivial resources, but we need to change this (see e.g. https://github.com/elastic/elasticsearch/pull/103356, https://github.com/elastic/elasticsearch/pull/103161, https://github.com/elastic/elasticsearch/pull/102778) and cannot do so correctly without pushing the right refcounting discipline through the security module. It would have been a bug if we hadn't noticed before 8.13 was released tho, see https://github.com/elastic/elasticsearch/pull/103953.

### PR DESCRIPTION
**Pull request by DaveCTurner**
`ActionFilter` implementations may be async, so we have to keep the
request alive while the chain is running.

Closes #103952

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/action/support/TransportActionFilterChainRefCountingTests.java [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+    }
+
+    static final ActionType<Response> TYPE = ActionType.localOnly("test:action");
+
+    public void testAsyncActionFilterRefCounting() {
+        final var countDownLatch = new CountDownLatch(2);
+        final var request = new Request();
+        try {
+            client().execute(TYPE, request, ActionListener.<Response>running(countDownLatch::countDown).delegateResponse((delegate, e) -> {
+                assertEquals("short-circuit failure", asInstanceOf(ElasticsearchException.class, e).getMessage());
```
Comment (jakelandis) [inline server/src/test/java/org/elasticsearch/action/support/TransportActionFilterChainRefCountingTests.java:51]: is this guaranteed to always be true ? (due to the random boolean below)

**Review by jakelandis — APPROVED** [review]
LGTM

File: server/src/test/java/org/elasticsearch/action/support/TransportActionFilterChainRefCountingTests.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/action/support/TransportActionFilterChainRefCountingTests.java:54]: Yes, it's ok if everything succeeds too, we're just checking that nothing else went wrong. I added a code comment in 84a194c1f94.

File: server/src/test/java/org/elasticsearch/action/support/TransportActionFilterChainRefCountingTests.java [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+            if (action.equals(TYPE.name())) {
+                randomFrom(EsExecutors.DIRECT_EXECUTOR_SERVICE, threadPool.generic()).execute(new AbstractRunnable() {
+                    @Override
+                    public void onFailure(Exception e) {
+                        fail(e);
+                    }
+
+                    @Override
+                    protected void doRun() {
+                        Thread.yield();
```
Comment (rjernst) [inline server/src/test/java/org/elasticsearch/action/support/TransportActionFilterChainRefCountingTests.java:123]: Is this meant to mimic work?

**Review by rjernst — APPROVED** [review]
LGTM

File: server/src/test/java/org/elasticsearch/action/support/TransportActionFilterChainRefCountingTests.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/action/support/TransportActionFilterChainRefCountingTests.java]: Kinda, but really I was just using it to make sure I had the right things covered by the tests and forgot to remove it. Gone now.

**Pull request merged**